Issue #39: Create an accessible modal to replace alerts

terchiem opened the pull request:
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description
- Installed `react-modal` to utilize a more accessible modal
- Created a reusable `CustomModal` component that can be used for alerts or confirmation modals
- Replaced all of the existing alerts and modals with `CustomModal`
- Refactored CSS class names around the modals

## Related Issue
Closes #39 
## Acceptance Criteria
- Find a library for an accessible modal component
- Create a reusable modal component 
- Replace existing alerts and dialogs with the modal component
<!-- Include AC from the Github issue -->

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|    | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|  ✓ | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before
![Screen Shot 2020-11-17 at 10 12 00 AM](https://user-images.githubusercontent.com/39573970/99429822-5c747c80-28bd-11eb-81b2-3ea6c7b127ae.png)

### After
![Screen Shot 2020-11-17 at 10 11 01 AM](https://user-images.githubusercontent.com/39573970/99429819-5bdbe600-28bd-11eb-8265-93fe826ee0c9.png)

## Testing Steps / QA Criteria
- Pull latest from `tc-gj-accessible-modal` and update dependancies with `npm install`
- Test out the modal in the following ways:
  - Joining a list with an invalid or no token
  - Successfully adding a new item to the list
  - Deleting an item from the list

alexandra-lim committed:
Merge pull request #41 from the-collab-lab/dev

Merge issues #11 + 12

GabbyJ committed:
Create accessible delete modal using react-modal.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Add aria-label description to delete modal.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Create Alert Modal and refactored for Join Existing List.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Refactor Add Item and Delete alerts with reusable modal.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

terchiem commented:
### Summary of work session (11/16/20)
- We looked around for packages that provided easy-to-use, accessible modals and landed on [react-modal](https://www.npmjs.com/package/react-modal)
- Refactored DeleteModal to use the more accessible react modal
- Created an AlertModal component to replace all alerts
  - replaced the alerts for entering an invalid token and for using a token that doesn't exist in JoinExistingList
  - replaced the alert for successful item deletion in List
  - replaced the alerts for successful item addition in AddItem*
- We tested the accessibility of the modals

### Todos
- Look into a better way to make the AlertModal reusable for DeleteModal as well

### * Challenges
Ideally, we wanted our modal alerts to be used the same way that the browser API's `alert` function behaves; using `alert('Item has been added!')` would display the modal with the appropriate message. We wrote a function which sets the appropriate states to display the modal:
```javascript
function setAlert(message) {
  setAlertModalMessage(message);
  setShowAlertModal(true);
}
```
The main challenge is that alerts were also being used in the  `addItem` function in the `listServices` file, which lives outside of React. One method we tried is to have `addItem` return a boolean for whether an item was successfully added or not, and then we would use `setAlert` with the appropriate message on the `AddItem` component in its `onSubmit` function. 

```javascript
const addItem = async (token, itemName, formData) => {
  db.collection('lists')
    .doc(token)
    .update({
      [itemName]: {
        name: formData.itemName,
        frequency: parseInt(formData.itemFrequency),
        lastPurchased: null,
      },
    })
    .then(() => {
      return true;
    })
    .catch((e) => {
      console.log('Error updating Firestore: ', e);
      return false;
    });
};
```
This unfortunately didn't work because it was always return `undefined` due to it being asynchronous code. We'll need to look into how to return it correctly. For now, we resorted to passing in the `setAlert` function to `addItem`.
```javascript
const addItem = async (token, itemName, formData, setAlert) => {
  db.collection('lists')
    .doc(token)
    .update({
      [itemName]: {
        name: formData.itemName,
        frequency: parseInt(formData.itemFrequency),
        lastPurchased: null,
      },
    })
    .then(() => {
      setAlert('Item has been submitted!');
    })
    .catch((e) => {
      console.log('Error updating Firestore: ', e);
      setAlert(
        `It isn't you, it's us. The item cannot be submitted at this time. Try again later while we look into it.`,
      );
    });
};
```

terchiem commented:
We managed to solve that problem we previously had thanks to [this post](https://dev.to/ivandotv/async-code-return-vs-return-await-m8) that Gabby found. Turns out that we were just missing a `return` on the promise:
```javascript
const addItem = async (token, itemName, formData) => {
  return db.collection('lists')
    .doc(token)
    .update({
      [itemName]: {
        name: formData.itemName,
        frequency: parseInt(formData.itemFrequency),
        lastPurchased: null,
      },
    })
    .then(() => {
      return true;
    })
    .catch((e) => {
      console.log('Error updating Firestore: ', e);
      return false;
    });
};
```

GabbyJ committed:
Refactor Add Item success confirmation to use async await.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Make universal modal that also works as DeleteModal.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Updated all components to use new universal modal.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Updated css for universal modal.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Add comments to functions in List.jsx.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

terchiem commented:
### Summary of work session (11/17/20)
- We removed the `DeleteModal` and `AlertModal` components made yesterday
- We created a `CustomModal` component that could be used universally for any alerts or confirmation windows
- We worked on replacing all of the alerts and modals to use `CustomModal`
- Refactored some CSS class names since we removed `DeleteModal`

### Challenges
The main challenge we faced today was in creating a universal modal that could be used to display an alert or be used to confirm an action. We settled on taking in a confirmation function as a component prop, and then adapting the modal if it was present:

```javascript
<div className="ModalButtons">
  <button className="Button cancelDelete" onClick={closeFunction}>
    {confirmFunction ? 'No' : 'Close'}
  </button>

  {confirmFunction && (
    <button
      className="Button PrimaryButton confirmDelete"
      onClick={confirmFunction}
    >
      Yes
    </button>
  )}
</div>
```
Our second challenge was to adapt our `deleteItem` function to be passed into the modal. We kept the confirmation function in state and when the user clicks an item's delete button, it would generate a new confirmation function for the item:
```javascript
//Sets up the modal for each item when user clicks the delete button
function deleteHandler(itemName) {
  setModalMessage(
    `Are you sure you want to delete "${itemName}" from the list?`,
  );
  setModalLabel(`Modal to confirm deletion of ${itemName}`);
  setConfirmFunction(createDeleteFunction(itemName));
  setModalIsOpen(true);
}

//Creates the delete function for the item to be used in the modal
function createDeleteFunction(itemName) {
  return function () {
    return function () {
      deleteItem(token, itemName);
      showAlert(`"${itemName}" has been deleted from your list.`);
    };
  };
}
```
Initially, `createDeleteFunction`  just returned a single function but we ran into a bug where the item would be deleted right away when the user clicked the delete button. Thanks to [this post](https://stackoverflow.com/questions/55621212/is-it-possible-to-react-usestate-in-react), we found out that giving a state setter a callback function will trigger it right away, so it had to be nested in another function that returns that function. 😅

terchiem committed:
Remove extra parameters from addItem function

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

edenbekele commented on src/Components/AddItem.jsx:
```diff
@@ -19,77 +27,98 @@ export default function AddItem({ token }) {
+        const successfulAdd = await addItem(token, sanitizedName, data);
+
+        if (successfulAdd) {
+          setAlert('Item has been submitted!');
```
When we tried to add an item after joining an existing list, the alert does not popup initially! We found that when we refresh our browser, however, it works. Here is the error message we received:

Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

![ezgif com-gif-maker](https://user-images.githubusercontent.com/61370161/99475487-4514ae00-2914-11eb-80fe-94a9f98202bf.gif)

terchiem replied:
Interesting! We'll look into it, thanks for catching this!

terchiem replied:
Thanks to EJ and Stacie, we found that this was happening because `onSnapshot` was being used when joining an existing list, which caused the app to redirect to `/list` every time the Firestore updated. This was fixed by switching `.onShapshot()` with `.get().then()` in `JoinExistingList.jsx`.
```javascript
// Before
db.collection('lists')
  .doc(lowerInput)
  .onSnapshot(function (querySnapshot) {
    // ...
  });

// After
db.collection('lists')
  .doc(lowerInput)
  .get()
  .then(function (querySnapshot) {
    // ...
  });
```

alexandra-lim commented on src/Components/CustomModal.jsx:
```diff
@@ -0,0 +1,38 @@
+      <p>{modalMessage}</p>
+      <div className="ModalButtons">
+        <button className="Button cancelDelete" onClick={closeFunction}>
+          {confirmFunction ? 'No' : 'Close'}
```
I like how you're both using a conditional here to change the text of the button.

edenbekele commented on src/styles/Modal.css:
```diff
@@ -1,13 +1,39 @@
-.Modal {
+.ModalOverlay {
```
Nice styling for the modal!

alexandra-lim commented on src/Components/CustomModal.jsx:
```diff
@@ -0,0 +1,38 @@
+          {confirmFunction ? 'No' : 'Close'}
+        </button>
+
+        {confirmFunction && (
```
Similar to the comment above, I really like how you've extended the functionality of the modal to conditionally render a "Yes" button if necessary

alexandra-lim commented on src/Components/List/List.jsx:
```diff
@@ -118,8 +118,42 @@ export default function List({ items, token }) {
+
+  function showAlert(message) {
+    setModalMessage(message);
+    setModalLabel('Alert');
```
We tested out the modal with Voiceover and it works really well! We did notice though that the screen reader just alerts the user to let them know that the window dialog is an alert, but doesn't specify what the alert message is. Maybe you could pass `message` into the `setModalLabel` function in addition to "Alert" to make it more specific?

terchiem replied:
We just noticed that the `aria-label` for the alerts that get displayed after a user confirms deletion weren't being read at all. This might be because we're just changing the alert message/label of the existing modal and not actually creating a new one.

Any thoughts @mxmason ? Would this be a case for ARIA live region? I know that you mentioned in the past that it is a little tricky to make it functional in React.

alexandra-lim commented:
Wonderful work this week, and really enjoyed reading the comments above regarding what roadblocks you encountered and how you tackled the challenges.

edenbekele reviewed (commented):
Great job on the code this week both of you, and for finishing your issue so quickly! Also amazingly thorough documenting of how you guys worked through the issue! 🤩 

We left one comment that is somewhat of a blocker in regards to the the alert that should pop up when submitting a new item.

terchiem committed:
Fix addItem bug by changing onSnapshot in JoinExistingList

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Fixed frequency null value by changing Submit name.
Updated 'item' to the specific item name.
Relocated states.
Looked into modal aria-labels.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Make items purchased once not be inactive but use the frequency instead.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

GabbyJ committed:
Fixed Add Item submit button on Firefox.
Updated Add Item success message.

Co-authored-by: Terry Chiem <terrychiem@gmail.com>
Co-authored-by: Gabrielle Jacobs <gabbyj424@gmail.com>

jmkuchman commented on src/Components/CustomModal.jsx:
```diff
@@ -0,0 +1,38 @@
+import '../styles/Modal.css';
+import '../styles/Button.css';
+
+export default function CustomModal({
```
I LOVE the thought and planning that has gone into this modal to make it useful for different use cases!

jmkuchman reviewed (approved):
Amazing work this week 👏 👏 I am so impressed with the refactoring and the changes this week. It's so great to see how this app has grown! You've taken existing functions and made them more uniform throughout the app and reused code effectively.

The app feels uniform throughout, and each of the modals functions as I'd expect, for both the "cancel" and "confirm" cases 👍 

I've also loved reading through the progress that you documented throughout the week! Reflecting on how your code is going and what you still need to accomplish is a practice that will help you so much, the more that you code. Nicely done, keep it up!
